### ISSUE TITLE
Dynamic nested fields do not obey schema type definition

### ISSUE THREAD

**Issue by juszzz** [typesense/typesense#973]
### Short Version
When defining a dynamic nested field in a schema with type `float`, instances of that dynamic nested field may instead have type `int64`, depending on whether or not the documents being inserted to the collection have field values that look like a float or an integer.

### Long Version
Consider documents that have the following `numericalSpecifications` field, which can contain different keys for different documents (e.g., one document might have an `Inside Height` key, while another might have `Load Capacity`):
```
    'numericalSpecifications': 
      {'Inside Height': {'units': 'inch', 'value': 2.25},
       'Inside Length': {'units': 'inch', 'value': 11.75},
       'Inside Width': {'units': 'inch', 'value': 10}
      },
```

The TS collection schema is defined as follows:
```
    "enable_nested_fields": True,
    "fields": [
        ...
        {"name": "numericalSpecifications", "type": "object", "index": True, "facet": True, "optional": True},
        {"name": "numericalSpecifications\..*\.units", "type": "string", "optional": True},
        {"name": "numericalSpecifications\..*\.value", "type": "float", "facet": True, "optional": True},
        ...]
```

After ingesting documents and retrieving the collection schema, there are `numericalSpecifications.foo.value` fields of type `int64` instead of `float`, as specified in the schema, e.g.:
```
  {'facet': True,
   'index': True,
   'infix': False,
   'locale': '',
   'name': 'numericalSpecifications.Inside Width.value',
   'optional': True,
   'sort': True,
   'type': 'int64'},
```

## Steps to reproduce

The following script will create a collection with a single document with `numericalSpecifications.foo.value` fields of differing types depending on whether the value in the document looks like a float or looks like an integer.
```
export TYPESENSE_API_KEY=xyz
export TYPESENSE_HOST=abc
​
curl "${TYPESENSE_HOST}/collections" \
       -X POST \
       -H "Content-Type: application/json" \
       -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" \
       -d '{
         "name": "repro_collection",
         "enable_nested_fields": true,
         "fields": [
           {"name": "id", "type": "string" },
           {"name": "numericalSpecifications", "type": "object", "index": true, "facet": true, "optional": true},
           {"name": "numericalSpecifications\\..*\\.units", "type": "string", "optional": true},
           {"name": "numericalSpecifications\\..*\\.value", "type": "float", "facet": true, "optional": true}
         ]
       }'
       
curl "${TYPESENSE_HOST}/collections/repro_collection/documents/import?action=create" \
        -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" \
        -H "Content-Type: text/plain" \
        -X POST \
        -d '{"id": "123", "numericalSpecifications": {"Load Capacity": {"units": "pound", "value": 56860.2}, "Overall Depth": {"units": "inch", "value": 42}, "Overall Height": {"units": "foot", "value": 12.5}}}'
​
curl "${TYPESENSE_HOST}/collections/repro_collection" \
        -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" | jq
```

## Expected Behavior

All `numericalSpecifications.foo.value` fields should be of type `float`, regardless of whether or not the first document of a given dynamic field has a value that looks like an integer or not.

## Actual Behavior

`numericalSpecifications.Load Capacity.value` and `numericalSpecifications.Overall Height.value` are of type `float`, but `numericalSpecifications.Overall Depth.value` is of type `int64`.

## Metadata

**Typesense Version**: 0.24.0

**OS**: MacOS Ventura 13.2.1

**Comment by juszzz** [comment]
Relevant [Slack thread](https://typesense-community.slack.com/archives/C01P749MET0/p1681246209579169)

**Comment by kishorenc** [comment]
@juszzz 

I've a fixed for this in `typesense/typesense:0.25.0.rc24`.

**Comment by vikjarmale** [comment]
> ## Description
> ### Short Version
> When defining a dynamic nested field in a schema with type `float`, instances of that dynamic nested field may instead have type `int64`, depending on whether or not the documents being inserted to the collection have field values that look like a float or an integer.
> 
> ### Long Version
> Consider documents that have the following `numericalSpecifications` field, which can contain different keys for different documents (e.g., one document might have an `Inside Height` key, while another might have `Load Capacity`):
> 
> ```
>     'numericalSpecifications': 
>       {'Inside Height': {'units': 'inch', 'value': 2.25},
>        'Inside Length': {'units': 'inch', 'value': 11.75},
>        'Inside Width': {'units': 'inch', 'value': 10}
>       },
> ```
> 
> The TS collection schema is defined as follows:
> 
> ```
>     "enable_nested_fields": True,
>     "fields": [
>         ...
>         {"name": "numericalSpecifications", "type": "object", "index": True, "facet": True, "optional": True},
>         {"name": "numericalSpecifications\..*\.units", "type": "string", "optional": True},
>         {"name": "numericalSpecifications\..*\.value", "type": "float", "facet": True, "optional": True},
>         ...]
> ```
> 
> After ingesting documents and retrieving the collection schema, there are `numericalSpecifications.foo.value` fields of type `int64` instead of `float`, as specified in the schema, e.g.:
> 
> ```
>   {'facet': True,
>    'index': True,
>    'infix': False,
>    'locale': '',
>    'name': 'numericalSpecifications.Inside Width.value',
>    'optional': True,
>    'sort': True,
>    'type': 'int64'},
> ```
> 
> ## Steps to reproduce
> The following script will create a collection with a single document with `numericalSpecifications.foo.value` fields of differing types depending on whether the value in the document looks like a float or looks like an integer.
> 
> ```
> export TYPESENSE_API_KEY=xyz
> export TYPESENSE_HOST=abc
> ​
> curl "${TYPESENSE_HOST}/collections" \
>        -X POST \
>        -H "Content-Type: application/json" \
>        -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" \
>        -d '{
>          "name": "repro_collection",
>          "enable_nested_fields": true,
>          "fields": [
>            {"name": "id", "type": "string" },
>            {"name": "numericalSpecifications", "type": "object", "index": true, "facet": true, "optional": true},
>            {"name": "numericalSpecifications\\..*\\.units", "type": "string", "optional": true},
>            {"name": "numericalSpecifications\\..*\\.value", "type": "float", "facet": true, "optional": true}
>          ]
>        }'
>        
> curl "${TYPESENSE_HOST}/collections/repro_collection/documents/import?action=create" \
>         -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" \
>         -H "Content-Type: text/plain" \
>         -X POST \
>         -d '{"id": "123", "numericalSpecifications": {"Load Capacity": {"units": "pound", "value": 56860.2}, "Overall Depth": {"units": "inch", "value": 42}, "Overall Height": {"units": "foot", "value": 12.5}}}'
> ​
> curl "${TYPESENSE_HOST}/collections/repro_collection" \
>         -H "X-TYPESENSE-API-KEY: ${TYPESENSE_API_KEY}" | jq
> ```
> 
> ## Expected Behavior
> All `numericalSpecifications.foo.value` fields should be of type `float`, regardless of whether or not the first document of a given dynamic field has a value that looks like an integer or not.
> 
> ## Actual Behavior
> `numericalSpecifications.Load Capacity.value` and `numericalSpecifications.Overall Height.value` are of type `float`, but `numericalSpecifications.Overall Depth.value` is of type `int64`.
> 
> ## Metadata
> **Typesense Version**: 0.24.0
> 
> **OS**: MacOS Ventura 13.2.1

Testing.
